doc/src/labelmap.rst

```diff
@@ -40,7 +40,7 @@ As explained on the Howto page, valid type labels can contain any
 alphanumeric character, but cannot start with a number.  They can also
 contain standard characters such as square brackets "[" and "]", dash
 "-", underscore "_", plus "+" and equals "=" signs.  Note that type
-labels cannot contain the comment symbol '#'.
+labels must be put in quotation marks if they contain the comment symbol '#'.
```

Comment:
relaxing this requirement may require significant changes to read_data and molecule template file parsing

Reply:
I have not seen any code enforcing this condition. So I am not relaxing anything. So far the only complication is that when processing input commands, "#' starts a comment and thus quotation marks need to be use (and will be removed). That makes it more a problem of the input processing than an issue of the label map processing. There should probably be some check for what is a valid label (restricting it to 7-bit ASCII for portability).

Reply:
FWIW, the following input seems to be working as expected:
```
region box block -2 2 -2 2 -2 2
create_box 4 box

labelmap atom 1 C1 2 C2 3 C3 4 "C#"

create_atoms 1 single 0.0 0.0 0.0
create_atoms 2 single 0.0 0.0 0.0
create_atoms 3 single 0.0 0.0 0.0
create_atoms 4 single 0.0 0.0 0.0

mass C1 1.0
mass C2 2.0
mass C3 3.0
mass "C#" 4.0

pair_style zero 2.0
pair_coeff * *
pair_coeff C1 C2
pair_coeff C3 "C#"

write_data label.data

clear

pair_style zero 3.0
read_data label.data

print "Atom labels: $(label(C1)) $(label(C2)) $(label(C3)) $(label(C#))"
```

I'll clarify in the docs that the quotation marks are only needed in a context with "#" would be interpreted as starting a comment, e.g. in the input file.

Reply:
type labels can also be used in the Atoms section of data files, directly replacing the numeric type (same with molecule templates). it looks like this is broken by # symbols, at the moment

Reply:
@jrgissing It seems that this breakage can be addressed with some rather small changes and without having to use quotes. See commit https://github.com/lammps/lammps/pull/2531/commits/d43051d07cde598c9289d7eab9643b24f367601e.
Do you recall to which other sections or parsers this would need to be propagated?
Adding support for quotes would be possible with using `utils::split_words()` instead of the Tokenizer class, but that is significantly slower and would slow down data file parsing, so I am hesitant to go this far.

In general, we should probably define a list of allowed characters (ideally limited to 7-bit ASCII, disallowing control characters and whitespace) and then set up a corresponding check against a regular expression with `utils::strmatch()`. So far, number, '#', and '*' are disallowed as first character, but by using quotes, the current code would also accept whitespace, which would be very difficult to support in data files.

Reply:
nice! other sections include the Bonds, Angles, Dihedrals and Impropers sections of data files, and the Types, Bonds, Angles, Dihedrals and Impropers sections of molecule templates

also, while we are trying to break things, atom types in several different force fields contain the single or double-quote character, such as the type c3"

Reply:
Supporting single/double quotes should be easier for anything using the Tokenizer class to parse lines of text into words.
In the input file those should just require using single/double quotes around the label.

However, this is something where we need some kind of well defined policy of what we want to allow or not, and suggestions for how to "encode" or "translate" unsupported characters etc. I have summarized my perspective on that in an email copied to all stakeholders. For something as fundamental as this feature, it is generally better to have a statement of the "we support X" kind rather than a "we do not support Y".

Reply:
Single/double quotes seem to be working with proper quoting in the input file and without changes to the read_data parser. The only thing that doesn't seem to work are "immediate" variables within an already quoted string (as with print and fix print), but that can be handled by defining explicit variables.
```
region box block -2 2 -2 2 -2 2
create_box 4 box

labelmap atom 1 C1 2 "C2'" 3 'C3"' 4 "C#"

create_atoms 1 single 0.0 0.0 0.0
create_atoms 2 single 0.0 0.0 0.0
create_atoms 3 single 0.0 0.0 0.0
create_atoms 4 single 0.0 0.0 0.0

mass C1 1.0
mass "C2'" 2.0
mass 'C3"' 3.0
mass "C#" 4.0

pair_style zero 2.0
pair_coeff * *
pair_coeff C1 "C2'"
pair_coeff 'C3"' "C#"

write_data label.data.new

clear

pair_style zero 3.0
read_data label.data.new

variable t1 equal label(C1)
variable t2 equal "label(C2')"
variable t3 equal 'label(C3")'
variable t4 equal "label(C#)"

print "Atom labels: ${t1} ${t2} ${t3} ${t4}"
```

Reply:
i wonder if this solution will run into problems for a bond type like C2'-C3"

support for escaping symbols might solve these issues and be easier to explain in the docs. but i don't have a sense of how reasonable it would be to implement

Reply:
> i wonder if this solution will run into problems for a bond type like C2'-C3"

You could use triple quotes `labelmap atom 1 C1 2 """C2'-C2" """ 3 'C3"' 4 "C#"`, but a blank after 'C2"' is currently required, so that in the label definition that trailing blank must be removed.

Reply:
Escaping would not be easy. Encoding might be an alternative. Like in HTML where a literal '<' becomes '&lt;'

Reply:
This now works, too:
```
region box block -2 2 -2 2 -2 2
create_box 4 box

labelmap atom 1 C1 2 """C2'-C2" """ 3 'C3"' 4 """ C1"-C2'"""
create_atoms 1 single 0.0 0.0 0.0
create_atoms 2 single 0.0 0.0 0.0
create_atoms 3 single 0.0 0.0 0.0
create_atoms 4 single 0.0 0.0 0.0

mass C1 1.0
mass """C2'-C2" """ 2.0
mass 'C3"' 3.0
mass """ C1"-C2'""" 4.0

pair_style zero 2.0
pair_coeff * *
pair_coeff C1 """C2'-C2" """
pair_coeff 'C3"' """ C1"-C2'"""

write_data label.data.new

clear

pair_style zero 3.0
read_data label.data.new

variable t1 equal label(C1)
variable t2 equal """label(C2'-C2")"""
variable t3 equal 'label(C3")'
variable t4 equal """label(C1"-C2')"""

print "Atom labels: ${t1} ${t2} ${t3} ${t4}"
```

Reply:
Looks good to me

Reply:
Using triple quotes with a blank before and after the label seems to be the most general way of quoting for all special cases so far. Just added a "labelmap write" command that should work similar to "write_coeffs" but for the label maps and used that there, too.